distributed/shuffle/shuffle_extension.py

```diff
@@ -0,0 +1,348 @@
+    id: ShuffleId
+    empty: pd.DataFrame
+    column: str
+    npartitions: int
+
+
+@dataclass(frozen=True, eq=False)
+class ShuffleMetadata(NewShuffleMetadata):
+    """
+    Metadata every worker needs to share about a shuffle.
+
+    A `ShuffleMetadata` is created with a task and sent to all workers
+    over the `ShuffleWorkerExtension.shuffle_init` RPC.
+    """
+
+    workers: list[str]
+
+    def worker_for(self, output_partition: int) -> str:
+        "Get the address of the worker which should hold this output partition number"
+        assert output_partition >= 0, f"Negative output partition: {output_partition}"
+        if output_partition >= self.npartitions:
+            raise IndexError(
+                f"Output partition {output_partition} does not exist in a shuffle producing {self.npartitions} partitions"
+            )
+        i = len(self.workers) * output_partition // self.npartitions
+        return self.workers[i]
+
+    def partition_range(self, worker: str) -> tuple[int, int]:
+        "Get the output partition numbers (inclusive) that a worker will hold"
+        i = self.workers.index(worker)
+        first = math.ceil(self.npartitions * i / len(self.workers))
+        last = math.ceil(self.npartitions * (i + 1) / len(self.workers)) - 1
+        return first, last
+
+    def npartitions_for(self, worker: str) -> int:
+        "Get the number of output partitions a worker will hold"
+        first, last = self.partition_range(worker)
+        return last - first + 1
+
+
+class Shuffle:
+    "State for a single active shuffle"
+
+    def __init__(self, metadata: ShuffleMetadata, worker: Worker) -> None:
+        self.metadata = metadata
+        self.worker = worker
+        self.output_partitions: defaultdict[int, list[pd.DataFrame]] = defaultdict(list)
+        self.output_partitions_left = metadata.npartitions_for(worker.address)
+        self.transferred = False
+
+    def receive(self, output_partition: int, data: pd.DataFrame) -> None:
+        assert not self.transferred, "`receive` called after barrier task"
+        self.output_partitions[output_partition].append(data)
+
+    async def add_partition(self, data: pd.DataFrame) -> None:
+        assert not self.transferred, "`add_partition` called after barrier task"
+        tasks = []
+        # TODO grouping is blocking, should it be offloaded to a thread?
+        # It mostly doesn't release the GIL though, so may not make much difference.
+        for output_partition, data in data.groupby(self.metadata.column):
+            addr = self.metadata.worker_for(int(output_partition))
+            task = asyncio.create_task(
+                self.worker.rpc(addr).shuffle_receive(
+                    shuffle_id=self.metadata.id,
+                    output_partition=output_partition,
+                    data=to_serialize(data),
+                )
+            )
+            tasks.append(task)
+
+        # TODO handle errors and cancellation here
```

Comment:
Add a comment that we need the rerun group to do it properly which is why we haven't implemented anything, yet